### PR TITLE
[NOSQUASH] Use cpp17, and some cleanups

### PR DESCRIPTION
**Pull request by Desour**
Descriptions for each commit:
* Bumps the mimimum gcc and clang versions to the minimum that ubuntu 20.04 supports (gcc 7 and clang 6). And re-enables the CI builds.
* Bumps the C++ std to 17. (Mostly supported by gcc 7 and clang 6, see <https://en.cppreference.com/w/cpp/compiler_support/17>.)
* Fixes a bug where `try_lock` is called without using the return value. (Found because of `[[nodiscard]]`.)
  Note: The Thread logic still looks faulty to me, i.e. I'm not sure if there's a race condition if the `try_lock` fails, or if we sometimes call `try_lock` on a mutex locked by the current thread (which is UB). Someone should fix or explain this at some point.
* Our `Optional` type is replaced with `std::optional`.
* Forward declarations of the form `namespace foo { namespace bar { ... }}` are replaced by the new syntax `namespace foo::bar { ... }`.
* Our `STATIC_ASSERT` macro is replaced by `static_assert`. (This is not C++17 specific, and could've been done much earlier.)
* The `-Wimplicit-fallthrough` warnings are no longer disabled. The `[[fallthrough]]` attribute is now guaranteed to exist. (Btw. we can easily use attributes of newer C++ standards, see also [feature testing](https://isocpp.org/std/standing-documents/sd-6-sg10-feature-test-recommendations#testing-for-the-presence-of-an-attribute-__has_cpp_attribute).)
  (Note: The same *could* be done for the unused-parameter warnings, with `[[maybe_unused]]`. But there seem to be many irrlicht functions triggering the warning, and it doesn't seem worth it.)
* `[[noreturn]]` is used, as opposed to the compiler-specific syntax. (Also not new in C++17.)

(Related: #13583)

Related (no dependency, but you'll get deprecation warnings without): https://github.com/minetest/irrlicht/pull/202

## To do

This PR is a Ready for Review.

## How to test

Build.

### CODE REVIEW

File: src/client/hud.cpp [hunk 1]
```diff
@@ -440,6 +440,7 @@ void Hud::drawLuaElements(const v3s16 &camera_offset)
 			case HUD_ELEM_IMAGE_WAYPOINT: {
 				if (!calculateScreenPos(camera_offset, e, &pos))
 					break;
+				[[fallthrough]];
```
Comment (lhofhansl) [inline src/client/hud.cpp:443]: This is a syntax I have not seen before.

Comment (Desour) [reply src/client/hud.cpp:443]: Nice parts of the C++ syntax be like: `[[]][](){}()`

File: src/rollback_interface.cpp [hunk 2]
```diff
@@ -65,6 +65,7 @@ std::string RollbackAction::toString() const
 		os << ", " << itos(n_new.param2);
 		os << ", " << serializeJsonString(n_new.meta);
 		os << ')';
+		break;
```
Comment (lhofhansl) [inline src/rollback_interface.cpp:68]: This is different behavior now. Is that intentional? If so it was bug before.

**Comment by lhofhansl** [comment]
Nice. Looks good. The one extra break looks weird, unless that was a (glaring) bug before.

**Review by sfan5 — APPROVED** [review]
LGTM otherwise

File: src/rollback_interface.cpp (hunk 2 above)
Comment (Desour) [reply src/rollback_interface.cpp:68]: It was an obvious bug.
Technically the behaviour of `l_rollback_revert_actions_by` changes a little, but I doubt mods will break due to this.

File: src/threading/thread.cpp [hunk 3]
```diff
@@ -95,11 +95,6 @@ Thread::~Thread()
 		wait();
 #endif
 	}
-
-	// Make sure start finished mutex is unlocked before it's destroyed
-	if (m_start_finished_mutex.try_lock())
-		m_start_finished_mutex.unlock();
```
[the remark this answers is not in the document]
Comment (Desour) [reply src/threading/thread.cpp:101]: The `std::unique_lock`s (=`MutexAutoLock`s) will always automatically release the lock. So it's not possible that the lock is held at this time (assuming the destructor isn't called while another thread uses the object).
But, thinking of it, I'm not sure if there isn't some weird scenario where that assumption is false, e.g. if someone starts the Thread from one thread while the Thread is destroyed by another thread.
I'll just add it back, just to be safe.

**Review by numberZero — COMMENTED** [review]
LGTM

**Pull request merged**